include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -34,25 +34,70 @@ namespace __par_backend_hetero
 {
 
 template <typename... _Name>
-class __reduce_seq_kernel;
+class __reduce_small_kernel;
 
 template <typename... _Name>
-class __reduce_small_kernel;
+class __reduce_mid_main_kernel;
+
+template <typename... _Name>
+class __reduce_mid_leaf_kernel;
 
-template <bool _IsGPU, typename... _Name>
+template <typename... _Name>
 class __reduce_kernel;
 
+// Single work group kernel that transforms and reduces __n elements to the single result.
+template <typename _Tp, typename _NDItemId, typename _Size, typename _TransformPattern, typename _ReducePattern,
+          typename _InitType, typename _AccLocal, typename _Res, typename... _Acc>
+void
+__work_group_reduce_kernel(const _NDItemId __item_id, const _Size __n, const _Size __n_items,
+                           const _TransformPattern __transform_pattern, const _ReducePattern __reduce_pattern,
+                           const _InitType __init, const _AccLocal& __local_mem, _Res& __res_acc, const _Acc&... __acc)
+{
+    auto __local_idx = __item_id.get_local_id(0);
+    // 1. Initialization (transform part). Fill local memory
+    __transform_pattern(__item_id, __n, /*global_offset*/ (_Size)0, __local_mem, __acc...);
+    __dpl_sycl::__group_barrier(__item_id);
+    // 2. Reduce within work group using local memory
+    _Tp __result = __reduce_pattern(__item_id, __n_items, __local_mem);
+    if (__local_idx == 0)
+    {
+        __reduce_pattern.apply_init(__init, __result);
+        __res_acc[0] = __result;
+    }
+}
+
+// Device kernel that transforms and reduces __n elements to the number of work groups preliminary results.
+template <typename _Tp, typename _NDItemId, typename _Size, typename _TransformPattern, typename _ReducePattern,
+          typename _InitType, typename _AccLocal, typename _Tmp, typename... _Acc>
+void
+__device_reduce_kernel(const _NDItemId __item_id, const _Size __n, const _Size __n_items,
+                       const _TransformPattern __transform_pattern, const _ReducePattern __reduce_pattern,
+                       const _InitType __init, const _AccLocal& __local_mem, _Tmp& __temp_acc, const _Acc&... __acc)
```

Comment:
Const is missing for `_Tmp& __temp_acc`, `__init` is not used.

Reply:
Applied.